### PR TITLE
Minor improvements

### PR DESCRIPTION
**Pull request by jdelauney**
- Adding missing css properties and some minor cosmetics change in the default template.
- Change menu template to use list html ul/li tags instead of table.
- Made little cosmetic changes and fix layout in pasdoc gui with Lazarus
- Update Jquery to ver3.5.1 
- Update TipueSearch to ver7.1 and fix integration

defaultcss1

defaultcss2

### CODE REVIEW

File: source/component/PasDoc_GenHtml.pas [hunk 1]
```diff
@@ -2496,18 +2496,19 @@ function TGenericHTMLDocGenerator.FormatTableOfContents(
...
   function MakeNavigation: string;
 
     function LocalMakeLink(const Filename, Caption: string): string;
     begin
-      Result := '<a href="' + EscapeURL(Filename) + '" class="navigation">' +
-        ConvertString(Caption) + '</a>';
+      Result := '<a href="' + EscapeURL(Filename) + '">' + ConvertString(Caption) + '</a>';
     end;
 
     function LocalMakeParagraphLink(const Filename, Caption: string): string; overload;
```
Comment (Bi0T1N) [inline source/component/PasDoc_GenHtml.pas:2509]: Should probably be named `LocalMakeListItemLink` now

**Comment by Bi0T1N** [comment]
Very good! 👍 
Should resolve #97 as well as your #76.

File: source/component/PasDoc_GenHtml.pas [hunk 2]
```diff
@@ -2520,33 +2521,49 @@ function THTMLDocGenerator.MakeBodyBegin: string;
     i: Integer;
   begin
     Result := '';
+    Skip := false;
 
-    if Title <> '' then
+    if Introduction <> nil then
```
Comment (Bi0T1N) [inline source/component/PasDoc_GenHtml.pas:2526]: I guess here applies the same which was mentioned in https://github.com/pasdoc/pasdoc/pull/95#pullrequestreview-417402790 and thus the `Skip` variable and additional code could be removed.

File: source/gui/frmhelpgeneratorunit.pas [hunk 3]
```diff
@@ -1574,7 +1574,8 @@ procedure TfrmHelpGenerator.MenuGenerateRunClick(Sender: TObject);
   { Switch to "Generate" page }
   lbNavigation.ItemIndex := lbNavigation.Items.IndexOfObject(pageGenerate);
   lbNavigationClick(nil);
-
+  Application.ProcessMessages;
+  Sleep(500);
   ButtonGenerateDocsClick(nil);
```
Comment (michaliskambi) [inline source/gui/frmhelpgeneratorunit.pas:1579]: You process messages (once), then hang for half of the second. Why is this necessary? 

If you have a case when this is needed, please add a comment describing it. Otherwise please revert  this change,

File: source/gui/frmhelpgeneratorunit.lfm [hunk 4]
```diff
@@ -1,18 +1,17 @@
 object frmHelpGenerator: TfrmHelpGenerator
```
Comment (michaliskambi) [inline source/gui/frmhelpgeneratorunit.lfm:1]: Can you describe what has changed, what I should look for when testing this change visually? 

You mention in PR "Made little cosmetic changes and fix layout in pasdoc gui with Lazarus" -- please expand this description :)

File: source/component/tipue/tipuesearch_set.js [hunk 5]
```diff
@@ -42,13 +43,15 @@ var tipuesearch_stem = {'words': [
...
 
-// Related searches
+// Related
 
-var tipuesearch_related = {'searches': [
-     // Not relevant for PasDoc docs: {'search': 'tipue', 'related': 'Tipue Search'},
-     // Not relevant for PasDoc docs: {'search': 'tipue', 'before': 'Tipue Search', 'related': 'Getting Started'},
-     // Not relevant for PasDoc docs: {'search': 'tipue', 'before': 'Tipue', 'related': 'jQuery'},
-     // Not relevant for PasDoc docs: {'search': 'tipue', 'before': 'Tipue', 'related': 'Blog'}
+var tipuesearch_related = {'Related': [
```
Comment (michaliskambi) [inline source/component/tipue/tipuesearch_set.js:48]: As the original comments here say: some of these lines are not relevant for PasDoc (having "related" results that lead to things not interesting to Pascal developers that read/search docs). Please remove these lines.

File: source/component/tipue/PasDoc_Tipue.pas [hunk 6]
```diff
@@ -244,12 +253,12 @@ procedure TipueAddFiles(Units: TPasUnits;
   end;
 
 const
-  TipueSearchCss: {$I tipuesearch.css.inc};
```
Comment (michaliskambi) [inline source/component/tipue/PasDoc_Tipue.pas:247]: Do not change the code formatting, esp. on lines where you didn't change anything :) In PasDoc we declare constants with types like

```pascal
const
  Name: Type = Value;
```

with no space between `Name` and `:`, and 1 space between `:` and `Type`.

The `Type = Value` in this case is defined in the include file.

The original code, as far as I see, was following our formatting. Please restore it.

File: source/component/pasdoc.css [hunk 7]
```diff
@@ -116,9 +169,34 @@ table.detail td.itemcode { width:100%; }
...
+
+dl.parameters dd:after{
+  display: block;
+  content: ';
+}
+
+code {
+  font-family: monospace;
+  font-size:1.2em;
+  font-style:  italic;
```
Comment (michaliskambi) [inline source/component/pasdoc.css:197]: I'll have to review some of these visual changes, I especially hesitate about this -- making code italic. I'll look at the resulting doc with new CSS later.

File: source/component/PasDoc_GenHtml.pas (hunk 2 above)
Comment (michaliskambi) [reply source/component/PasDoc_GenHtml.pas:2526]: I second that :) Just link to `index.html` always, and `Skip` should not be needed.

File: source/component/PasDoc_GenHtml.pas (hunk 1 above)
Comment (michaliskambi) [reply source/component/PasDoc_GenHtml.pas:2509]: I second that :)

Comment (jdelauney) [reply source/component/PasDoc_GenHtml.pas:2509]: I'm agree with that

File: source/component/PasDoc_GenHtml.pas (hunk 2 above)
Comment (jdelauney) [reply source/component/PasDoc_GenHtml.pas:2526]: ok

File: source/gui/frmhelpgeneratorunit.pas (hunk 3 above)
Comment (jdelauney) [reply source/gui/frmhelpgeneratorunit.pas:1579]: Application.processmessages force to send the messages to the system and sleep let the the time to those messages to be treated and the to ui to be repaint correctly. Generating is huge and long process. Without that the ui is not refresh correctly, because messages are in and stay in queue until the generating process have finish.

File: source/gui/frmhelpgeneratorunit.lfm (hunk 4 above)
Comment (jdelauney) [reply source/gui/frmhelpgeneratorunit.lfm:1]: Just some minor change,  i fix some component to alCustom instead alNone and fix the with for prevent the automatic resizing of the form

File: source/component/tipue/tipuesearch_set.js (hunk 5 above)
Comment (jdelauney) [reply source/component/tipue/tipuesearch_set.js:48]: I'm not take a look inside the js, just download the latest version and convert it with the tools in pasdoc

File: source/component/tipue/PasDoc_Tipue.pas (hunk 6 above)
Comment (jdelauney) [reply source/component/tipue/PasDoc_Tipue.pas:247]: Sorry it's my habit, i'm not take care

File: source/component/pasdoc.css (hunk 7 above)
Comment (jdelauney) [reply source/component/pasdoc.css:197]: font-italic Is not necessary like you can see on the screenshot

File: source/component/PasDoc_GenHtml.pas (hunk 1 above)
Comment (jdelauney) [reply source/component/PasDoc_GenHtml.pas:2509]: You talk about the  function LocalMakeParagraphLink ??

**Comment by jdelauney** [comment]
I've amend the last commit

File: source/component/PasDoc_GenHtml.pas [hunk 8]
```diff
@@ -2521,65 +2520,78 @@ function THTMLDocGenerator.MakeBodyBegin: string;
...
-      if Introduction.ShortTitle = '' then
-        Result := Result + LocalMakeParagraphLink(Introduction.OutputFileName, trIntroduction) else
-        Result := Result + LocalMakeParagraphLink(Introduction.OutputFileName, Introduction.ShortTitle);
+      if (Introduction <> nil) then
+      begin
+        if Introduction.ShortTitle='' then
+          Result := Result + '<h2>' + LocalMakeLink('index.html', Introduction.Title) + '</h2>'
+        else
+          Result := Result + '<h2>' + LocalMakeLink('index.html', Introduction.ShortTitle) + '</h2>';
+      end;
```
Comment (michaliskambi) [inline source/component/PasDoc_GenHtml.pas:2535]: You changed too much here, I'm afraid. The previous review comments requested to link to `index.html` if you just want to link to the "home page". But an explicit link to "introduction" should still look like in the original, i.e. point to `Introduction.OutputFileName`.

Also, keep the formatting of the original code and don't make changes when you didn't want. Looking at this change:

- you changed `if Introduction <> nil then` to `if (Introduction <> nil) then`, we don't use such excessive parenthesis
- you removed spaces from `Introduction.ShortTitle = ''` comparison
- you left a comment `//if (Title = '') and` below

Please avoid changing code formatting -- we try to follow consistent Pascal coding standards in PasDoc. We're definitely not ideal (a lot of people worked on this project and there are pieces of code with inconsistent lines), but we try :)

Please clean this change a little -- you want to just change `LocalMakeParagraphLink` to `LocalMakeLink` as far as I see. As as I see no other change in code in case `Introduction <> nil` should be needed.

File: source/component/pasdoc.css (hunk 7 above)
Comment (michaliskambi) [reply source/component/pasdoc.css:197]: Please remove the `font-italic` then.

File: source/component/tipue/tipuesearch_set.js (hunk 5 above)
Comment (michaliskambi) [reply source/component/tipue/tipuesearch_set.js:47]: You still left 
`{'search': 'tipue', 'related': 'Search', 'include': 1}`
and
`{'search': 'tipue', 'related': 'Support'}`.

If you want to leave them, it may be OK -- but add a comment why :) We want to have here only stuff really interesting to readers of a particular API documentation. These readers are not interested in Tipue, they also didn't develop the Tipue+PasDoc integration, they just used "search" in some API docs they found on the Internet.

File: source/gui/frmhelpgeneratorunit.pas (hunk 3 above)
Comment (michaliskambi) [reply source/gui/frmhelpgeneratorunit.pas:1579]: Thanks for explaining! I get the point of `Application.ProcessMessages`.

I still don't think `Sleep(500)` is useful. You wrote above """sleep let the the time to those messages to be treated and the to ui to be repaint correctly""" but that's not what the `Sleep` routine does. It merely hangs the process for half a second, nothing gets repainted during this time.

Can you test whether the `Sleep(500)` is really useful to help repaint? And on what platform?

File: source/gui/frmhelpgeneratorunit.lfm (hunk 4 above)
Comment (jdelauney) [reply source/gui/frmhelpgeneratorunit.lfm:1]: Hi, no problem, i'm busy too, i'll make the changes soon as possible. Surely this week-end ;)

Cheers

File: source/component/pasdoc.css (hunk 7 above)
Comment (jdelauney) [reply source/component/pasdoc.css]: Removed

File: source/component/PasDoc_GenHtml.pas (hunk 8 above)
Comment (jdelauney) [reply source/component/PasDoc_GenHtml.pas:2535]: Fixed

File: source/gui/frmhelpgeneratorunit.pas (hunk 3 above)
Comment (jdelauney) [reply source/gui/frmhelpgeneratorunit.pas:1579]: I removed the sleep, it seems not necessary, during my tests

File: source/component/tipue/tipuesearch_set.js (hunk 5 above)
Comment (jdelauney) [reply source/component/tipue/tipuesearch_set.js:47]: Added "// Not relevant for PasDoc" comment  in sections in file  tipuesearch_set.js

**Comment by michaliskambi** [comment]
@jdelauney Thank you! I'll  review it most likely at next weekend (I'm a bit swamped now with a work around [X3D / Castle Game Engine that I want to finish 3rd July](https://github.com/michaliskambi/x3d-tests/wiki/X3D-version-4:-New-features-of-materials,-lights-and-textures) :) ).

**Comment by michaliskambi** [comment]
Sorry for delay! I *do* have this on my TODO list and will apply soon :)

**Comment by Bi0T1N** [comment]
> Sorry for delay! I _do_ have this on my TODO list and will apply soon :)

@michaliskambi Take this as a reminder ;)

**Review by michaliskambi — APPROVED** [review]

**Comment by michaliskambi** [comment]
Sorry for such delay!

Merged. Thank you very much!

Please in the future make pull requests only for one specific feature. These are easier to review and thus apply :)

1. I applied a number of fixes to the CSS.

    Your CSS changes were a mixture of things we want (e.g. divs for columns), stuff that is nice (some navigation changes), stuff that we don't want (broken look of search button on narrow screens, inconsistent tipue results look) and stuff that is a matter of taste (like color changes to navigation and tables).

2. I had to revert your changes to pasdoc_gui, I'm afraid. As far as I can see you removed all anchors, and moved many controls assuming that things have exactly the same size in pixels always. This unfortunately means that the GUI looks bad for me (and I guess for others too, if they have different OS theme than you use).

    See Lazarus documentation about how to properly anchor controls and use auto-sizing to make GUI look correct under all themes/font sizes: https://wiki.freepascal.org/Anchor_Sides , https://wiki.freepascal.org/Autosize_/_Layout .

See the screenshots from the pasdoc_gui after your changes, showing the problems:

b1
b2
b3
b4

Here are the same views in correct version (to which I reverted the lfm):

g1
g2
g3
g4
